plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go

```
@@ -0,0 +1,532 @@
+	Name   string
+	Origin string
+	Path   string
+
+	// Subscription mode and interval
+	SubscriptionMode string            `toml:"subscription_mode"`
+	SampleInterval   internal.Duration `toml:"sample_interval"`
+
+	// Duplicate suppression
+	SuppressRedundant bool              `toml:"suppress_redundant"`
+	HeartbeatInterval internal.Duration `toml:"heartbeat_interval"`
+}
+
+// Start the http listener service
+func (c *CiscoTelemetryGNMI) Start(acc telegraf.Accumulator) error {
+	var err error
+	var ctx context.Context
+	var tlscfg *tls.Config
+	var request *gnmi.SubscribeRequest
+	c.acc = acc
+	ctx, c.cancel = context.WithCancel(context.Background())
+
+	// Validate configuration
+	if request, err = c.newSubscribeRequest(); err != nil {
+		return err
+	} else if c.Redial.Duration.Nanoseconds() <= 0 {
+		return fmt.Errorf("redial duration must be positive")
+	}
+
+	// Parse TLS config
+	if c.EnableTLS {
+		if tlscfg, err = c.ClientConfig.TLSConfig(); err != nil {
+			return err
+		}
+	}
+
+	if len(c.Username) > 0 {
+		ctx = metadata.AppendToOutgoingContext(ctx, "username", c.Username, "password", c.Password)
+	}
+
+	// Invert explicit alias list and prefill subscription names
+	c.aliases = make(map[string]string, len(c.Subscriptions)+len(c.Aliases))
+	for _, subscription := range c.Subscriptions {
+		path := subscription.Path
+		if len(subscription.Origin) > 0 {
+			path = subscription.Origin + ":" + path
+		}
+
+		name := subscription.Name
+		if len(name) == 0 {
+			name = path[strings.LastIndexByte(path, '/')+1:]
+		}
+		if len(name) > 0 {
+			c.aliases[path] = name
+		}
+	}
+	for alias, path := range c.Aliases {
+		c.aliases[path] = alias
+	}
+
+	// Create a goroutine for each device, dial and subscribe
+	c.wg.Add(len(c.Addresses))
+	for _, addr := range c.Addresses {
+		go func(address string) {
+			defer c.wg.Done()
+			for ctx.Err() == nil {
+				if err := c.subscribeGNMI(ctx, address, tlscfg, request); err != nil {
+					acc.AddError(err)
```

I noticed that I can get an error when shutting down Telegraf (since I don't have a real device to connect to):
```
E! [inputs.cisco_telemetry_gnmi]: Error in plugin: failed to setup subscription: rpc error: code = Canceled desc = context canceled
```

Also, can you make sure to double check for errors and for successful reconnecting if we lose connection to the monitored device.

---

Suppressing errors now when the context is cancelled. As for reconnecting we are running the same steps as for the first connection over and over again (with the redial timeout inbetween) until the connection succeeds.

---

:+1: Thanks, I see/remember the reconnect logic now